diesel/src/query_builder/select_by_statement/dsl_impls.rs

```diff
@@ -0,0 +1,365 @@
+where
+    STMT: FindDsl<PK>,
+{
+    type Output = SelectByStatement<S, STMT::Output>;
+
+    fn find(self, id: PK) -> Self::Output {
+        SelectByStatement::new(self.inner.find(id))
+    }
+}
+
+impl<ST, S, STMT, Expr> OrderDsl<Expr> for SelectByStatement<S, STMT>
+where
+    Expr: Expression,
+    Self: SelectQuery<SqlType = ST>,
+    STMT: OrderDsl<Expr>,
+    SelectByStatement<S, STMT::Output>: SelectQuery<SqlType = ST>,
+{
+    type Output = SelectByStatement<S, STMT::Output>;
+
+    fn order(self, expr: Expr) -> Self::Output {
+        SelectByStatement::new(self.inner.order(expr))
+    }
+}
+
+impl<ST, S, STMT, Expr> ThenOrderDsl<Expr> for SelectByStatement<S, STMT>
+where
+    Expr: Expression,
+    Self: SelectQuery<SqlType = ST>,
+    STMT: ThenOrderDsl<Expr>,
+    SelectByStatement<S, STMT::Output>: SelectQuery<SqlType = ST>,
+{
+    type Output = SelectByStatement<S, STMT::Output>;
+
+    fn then_order_by(self, expr: Expr) -> Self::Output {
+        SelectByStatement::new(self.inner.then_order_by(expr))
+    }
+}
+
+impl<ST, S, STMT> LimitDsl for SelectByStatement<S, STMT>
+where
+    Self: SelectQuery<SqlType = ST>,
+    STMT: LimitDsl,
+    SelectByStatement<S, STMT::Output>: SelectQuery<SqlType = ST>,
+{
+    type Output = SelectByStatement<S, STMT::Output>;
+
+    fn limit(self, limit: i64) -> Self::Output {
+        SelectByStatement::new(self.inner.limit(limit))
+    }
+}
+
+impl<ST, S, STMT> OffsetDsl for SelectByStatement<S, STMT>
+where
+    Self: SelectQuery<SqlType = ST>,
+    STMT: OffsetDsl,
+    SelectByStatement<S, STMT::Output>: SelectQuery<SqlType = ST>,
+{
+    type Output = SelectByStatement<S, STMT::Output>;
+
+    fn offset(self, offset: i64) -> Self::Output {
+        SelectByStatement::new(self.inner.offset(offset))
+    }
+}
+
+impl<S, STMT, Expr> GroupByDsl<Expr> for SelectByStatement<S, STMT>
+where
+    Self: SelectQuery,
+    Expr: Expression,
+    STMT: GroupByDsl<Expr>,
+{
+    type Output = STMT::Output;
+
+    fn group_by(self, expr: Expr) -> Self::Output {
+        self.inner.group_by(expr)
+    }
+}
+
+impl<ST, S, STMT, Lock> LockingDsl<Lock> for SelectByStatement<S, STMT>
+where
+    Self: SelectQuery<SqlType = ST>,
+    STMT: LockingDsl<Lock>,
+    SelectByStatement<S, STMT::Output>: SelectQuery<SqlType = ST>,
+{
+    type Output = SelectByStatement<S, STMT::Output>;
+
+    fn with_lock(self, lock: Lock) -> Self::Output {
+        SelectByStatement::new(self.inner.with_lock(lock))
+    }
+}
+
+impl<ST, S, STMT, Modifier> ModifyLockDsl<Modifier> for SelectByStatement<S, STMT>
+where
+    Self: SelectQuery,
+    STMT: ModifyLockDsl<Modifier>,
+    SelectByStatement<S, STMT::Output>: SelectQuery<SqlType = ST>,
+{
+    type Output = SelectByStatement<S, STMT::Output>;
+
+    fn modify_lock(self, modifier: Modifier) -> Self::Output {
+        SelectByStatement::new(self.inner.modify_lock(modifier))
+    }
+}
+
+// impl<'a, F, S, D, W, O, L, Of, G, DB> BoxedDsl<'a, DB>
```

Comment:
Any reason this is not implemented? I would just assume that this returns something like `SelectByStatement<S, BoxedSelectStatement<'a, S::Columns::SqlType, F, DB>>`

Reply:
Could you explain a little more about what BoxedDsl is and why we need it?
In general I'd like not to restrict `Stmt` to `SelectStatement`, unless there's a good reason.

Reply:
`BoxedDsl` is a way to concert a static query ast into something that could be chaged dynamically, i.e. where you can conditionally append a filer or something like that. Given the normal `SelectStatement` this is not possible, because each `QueryDsl` method return a new slightly changed type. For a `BoxedSelectStatement` this is different, as most methods will not change the type returned by such a method. This comes at a certain cost:
- `BoxedSelectstatement` requires a heap allocation per `QueryDsl` method call, whereas the normal `SelectStatement` uses zero sized static types in the most cases
- Building the corresponding sql will require a dynamic dispatch per query clause for `BoxedSelectStatements`, whereas it is mostly done at compile time for normal `SelectStatement`
- For normal `SelectStatement` we can do quite clever static compile time statement cashing based on the type ids, as for some/most cases the SQL will not change without a type change. This allows to skip building the sql at all if the query is already in the statement cache. `BoxedSelectStatement` cannot do that, as the same type could represent different sql. Therefore it needs to build the sql in all cases, as it is used as key into the statement cache. 

So for implementing `BoxedDsl` only for `Stmt` = `SelectStatement`: There are just no other cases in diesel that are useful, but we can probably also go with an generic impl here if possible.